cpp/src/detail/utility_wrappers.cu

```diff
@@ -110,5 +110,42 @@ void normalize(rmm::cuda_stream_view const& stream_view, value_t* d_value, size_
 template void normalize(rmm::cuda_stream_view const& stream_view, float* d_value, size_t size);
 template void normalize(rmm::cuda_stream_view const& stream_view, double* d_value, size_t size);
 
+template <typename vertex_t, typename edge_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<edge_t>> filter_degree_0_vertices(
+  raft::handle_t const& handle,
+  rmm::device_uvector<vertex_t>&& d_vertices,
+  rmm::device_uvector<edge_t>&& d_out_degs)
+{
+  auto zip_iter =
+    thrust::make_zip_iterator(thrust::make_tuple(d_vertices.begin(), d_out_degs.begin()));
+  auto zip_iter_end =
+    thrust::copy_if(handle.get_thrust_policy(),
+                    zip_iter,
+                    zip_iter + d_vertices.size(),
+                    zip_iter,
+                    [] __device__(auto pair) { return thrust::get<1>(pair) > 0; });
```

Comment:
FYI: https://github.com/NVIDIA/thrust/issues/1302
Maybe do copy_if in chunks or add check for d_vertices.size() and throw an exception if d_vertices.size() overflows 32 bit integer (if you expect this will unlikely to happen and we'd better wait for thrust folks to fix this).

Reply:
And I guess `thrust::remove_if` is more intuitive than `copy_if` here (unless you willing to copy in chunks). You may look for https://github.com/rapidsai/cugraph/pull/2253/files#diff-ce8c8b8ffdc670a97313ca4ce20de7bf8a18daa81f5a1fde50f3b162bf75b75bR1238
```
#if 1  // FIXME: work-around for the 32 bit integer overflow issue in thrust::remove,
       // thrust::remove_if, and thrust::copy_if (https://github.com/NVIDIA/thrust/issues/1302)
    rmm::device_uvector<vertex_t> tmp_indices(
      thrust::count_if(handle.get_thrust_policy(),
                       nbr_intersection_indices.begin(),
                       nbr_intersection_indices.end(),
                       detail::not_equal_t<vertex_t>{invalid_vertex_id<vertex_t>::value}),
      handle.get_stream());
    size_t num_copied{0};
    size_t num_scanned{0};
    while (num_scanned < nbr_intersection_indices.size()) {
      size_t this_scan_size = std::min(
        size_t{1} << 30,
        static_cast<size_t>(thrust::distance(nbr_intersection_indices.begin() + num_scanned,
                                             nbr_intersection_indices.end())));
      num_copied += static_cast<size_t>(thrust::distance(
        tmp_indices.begin() + num_copied,
        thrust::copy_if(handle.get_thrust_policy(),
                        nbr_intersection_indices.begin() + num_scanned,
                        nbr_intersection_indices.begin() + num_scanned + this_scan_size,
                        tmp_indices.begin() + num_copied,
                        detail::not_equal_t<vertex_t>{invalid_vertex_id<vertex_t>::value})));
      num_scanned += this_scan_size;
    }
    nbr_intersection_indices = std::move(tmp_indices);
#else
    nbr_intersection_indices.resize(
      thrust::distance(nbr_intersection_indices.begin(),
                       thrust::remove(handle.get_thrust_policy(),
                                      nbr_intersection_indices.begin(),
                                      nbr_intersection_indices.end(),
                                      invalid_vertex_id<vertex_t>::value)),
      handle.get_stream());
#endif
```

Reply:
Switched to remove_if.

Seems unlikely to have an overflow issue, at least with current memory sizes, as the number of elements in a vertex array on each partition is likely to be < 2^31-1.  But I added a FIXME so we can remember.

Reply:
Maybe add `CUGRAPH_EXPECTS(d_vertices.size() < std::numeric_limit<int32_t>::max())` as well. I agree that this is unlikely to happen, but if this happens in user side or large scale testing, it is very difficult for us to figure out this is actually due to the overflow. With the check, it will be way easier to figure out what went awry.

Reply:
Added the `CUGRAPH_EXPECTS` here and the other two places where I call `remove_if`.